### DISCUSSION TITLE
Composable screenshots with SideEffects

### DISCUSSION THREAD

**Discussion by St4B** [cashapp/paparazzi#500] [Ideas]
I am not sure whether I use `Paparazzi` wrongly or if there is a way to achieve the desired behavior, but I think that the API should provide a way to take snapshots after all side effects are executed. For example, I have a composable that renders differently when it has focus. In order to set focus programmatically in our screenshot test, we need to use `SideEffect` like that:

```kotlin
val focusRequester = remember { FocusRequester() }
MyAwesomeComposable(modifier = Modifier.focusRequester(focusRequester))
SideEffect { focusRequester.requestFocus() }
```

If I try to take a screenshot while the composable has focus:
```kotlin
paparazzi.snapshot("Awesome composable with focus") {
    val focusRequester = remember { FocusRequester() }
    MyAwesomeComposable(modifier = Modifier.focusRequester(focusRequester))
    SideEffect { focusRequester.requestFocus() }
}
```
then it seems that the screenshot was taken before the focus was gained (which I think is expected).

Maybe we need something like `IdlingResource`, or something to force all recompositions to happen.

**Comment by geoff-powell** [comment]
Can you try this with the latest version 1.3.5. We added the ability to rerender the first frame if a pending recomposition is detected.

**Reply by aileene-willow** [reply]
Updated to v1.3.5 but it's still not capturing focused state on a button component that's supposed to have an alpha on focus.

**Reply by St4B** [reply]
Hmm! It seems to work, but not in every case. For example, when I try to test a `TextField` with focus, the final state isn't fully rendered. The cursor is visible and the border is highlighted, but the label doesn't animate to the top. I suppose this happens because only the first frame is being rendered.
